text/1015-full-text-search.rst

```diff
@@ -58,74 +58,266 @@ Specification
 =============
 
 There's a lot of functionality involved in supporting FTS, but it's a fairly
-niche thing, so it makes sense to create a ``module fts`` for all of the FTS
-tools.
+niche thing, so it makes sense to create a ``module std::fts`` for all of the
+FTS tools.
 
-In EdgeDB we should avoid having an explicit "document" scalar like
-``tsvector``. Instead we can rely on various parameters, annotations and
-indexes to specify how the plain text ``str`` should be parsed into tokens.
-This will allow us to keep the queries involving FTS mostly implementation
-agnostic. Some FTS backends may have functions specific to them, but the goal
-is to provide good coverage of common FTS functionality.
+
+Indexing
+--------
+
+Indexes can be used to declare which properties should be subject to FTS
+functionality. They can also be used to annotate properties with search
+groups, language information, etc. Search groups can then be used to fine tune
+results by assigning different weights to each group during searches. Language
+information would determine what data is targeted based on the language of the
+search query. The proposed index definitions would thus have a semantic effect
+determining how exactly FTS should be carried out for a given object type.
+
+In EdgeDB we don't need an explicit "document" scalar that users interact
+with, because we don't allow arbitrary strings to be processed by FTS
+functions, only indexed properties. However, we still need a type similar to
+``tsvector`` to record things like weight category and language that
+correspond to a particular part of the indexed string. This type should be
+opaque and not valid as a result of a user query. Instead it can be used in
+index expressions and returned by functions that can only be used in index
+expressions. The type being opaque makes it possible to add more FTS functions
+that affect how the string is interpreted without having to change all the
+signatures.
+
+This processed document type and a couple of functions that use it are::
+
+  type fts::DocVector {
+    # This is hidden implementation, so it's more of a concept
```

Comment:
What does it mean "hidden"? Will it be opaque but still available? If it's opaque, how will you deal with it lacking introspection?

Reply:
I guess "index" functions won't be introspectable and that kinda solves this issue?

Reply:
I've added more explanation regarding the type. Any function works with this type should also be excluded from introspection.